bigtop-packages/src/charm/hadoop/layer-hadoop-plugin/reactive/apache_bigtop_plugin.py

```diff
@@ -0,0 +1,149 @@
+
+
+@when('apache-bigtop-plugin.hdfs.installed')
+@when('hadoop-plugin.joined', 'namenode.ready')
+@when_not('apache-bigtop-plugin.hdfs.ready')
+def send_principal_hdfs_info(principal, namenode):
+    """Send HDFS data when the namenode becomes ready."""
+    principal.set_installed(get_hadoop_version())
+    principal.set_hdfs_ready(namenode.namenodes(), namenode.port())
+    set_state('apache-bigtop-plugin.hdfs.ready')
+
+
+@when('apache-bigtop-plugin.hdfs.ready')
+@when('hadoop-plugin.joined')
+@when_not('namenode.ready')
+def clear_hdfs_ready(principal):
+    principal.clear_hdfs_ready()
+    remove_state('apache-bigtop-plugin.hdfs.ready')
+    remove_state('apache-bigtop-plugin.hdfs.installed')
+
+
+@when('bigtop.available', 'hadoop-plugin.joined', 'namenode.joined', 'resourcemanager.joined')
+@when_not('apache-bigtop-plugin.yarn.installed')
+def install_hadoop_client_yarn(principal, namenode, resourcemanager):
+    if namenode.namenodes() and resourcemanager.resourcemanagers():
+        hookenv.status_set('maintenance', 'installing plugin (yarn)')
+        nn_host = namenode.namenodes()[0]
+        rm_host = resourcemanager.resourcemanagers()[0]
+        bigtop = Bigtop()
+        hosts = {'namenode': nn_host, 'resourcemanager': rm_host}
+        bigtop.render_site_yaml(hosts=hosts, roles='hadoop-client')
+        bigtop.trigger_puppet()
+        set_state('apache-bigtop-plugin.yarn.installed')
+        hookenv.status_set('active', 'ready (HDFS & YARN)')
+    else:
+        hookenv.status_set('waiting', 'waiting for master fqdns')
+
+
+@when('apache-bigtop-plugin.yarn.installed')
+@when('hadoop-plugin.joined', 'resourcemanager.joined')
+@when_not('resourcemanager.ready')
+def send_rm_spec(principal, resourcemanager):
+    """Send our plugin spec so the resourcemanager can become ready."""
+    bigtop = Bigtop()
+    resourcemanager.set_local_spec(bigtop.spec())
+
+
+@when('apache-bigtop-plugin.yarn.installed')
+@when('hadoop-plugin.joined', 'resourcemanager.ready')
+@when_not('apache-bigtop-plugin.yarn.ready')
+def send_principal_yarn_info(principal, resourcemanager):
+    """Send YARN data when the resourcemanager becomes ready."""
+    version = get_hadoop_version()
+    principal.set_installed(version)
+    principal.set_yarn_ready(
+        resourcemanager.resourcemanagers(), resourcemanager.port(),
+        resourcemanager.hs_http(), resourcemanager.hs_ipc())
+    set_state('apache-bigtop-plugin.yarn.ready')
+
+
+@when('apache-bigtop-plugin.yarn.ready')
+@when('hadoop-plugin.joined')
+@when_not('resourcemanager.ready')
+def clear_yarn_ready(principal):
+    principal.clear_yarn_ready()
+    remove_state('apache-bigtop-plugin.yarn.ready')
+    remove_state('apache-bigtop-plugin.yarn.installed')
+
+
+@when_any('apache-bigtop-plugin.hdfs.installed', 'apache-bigtop-plugin.yarn.installed')
+@when('hadoop-plugin.joined')
+@when_none('namenode.spec.mismatch', 'resourcemanager.spec.mismatch')
+def update_status(principal):
```

Comment:
This routine feels a bit clunky to me. It seems like most of these things should be set as part of other routines.

I'm guessing that it's not that simple, but I always am sad to see a big block o' switch statements.